kinesis/src/main/scala/akka/stream/alpakka/kinesis/impl/ShardProcessor.scala

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright (C) 2016-2019 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.stream.alpakka.kinesis.impl
+
+import java.util.concurrent.Semaphore
+
+import akka.annotation.InternalApi
+import akka.stream.alpakka.kinesis.CommittableRecord
+import software.amazon.kinesis.lifecycle.ShutdownReason
+import software.amazon.kinesis.lifecycle.events._
+import software.amazon.kinesis.processor.ShardRecordProcessor
+import software.amazon.kinesis.retrieval.KinesisClientRecord
+import software.amazon.kinesis.retrieval.kpl.ExtendedSequenceNumber
+
+import scala.collection.JavaConverters._
+
+@InternalApi
+private[kinesis] class ShardProcessor(
+    callback: CommittableRecord => Unit
+) extends ShardRecordProcessor {
+
+  private var shardId: String = _
+  private var extendedSequenceNumber: ExtendedSequenceNumber = _
+
+  private val semaphore = new Semaphore(1)
+
+  var shutdown: Option[ShutdownReason] = None
+
+  override def initialize(initializationInput: InitializationInput): Unit = {
+    shardId = initializationInput.shardId()
+    extendedSequenceNumber = initializationInput.extendedSequenceNumber()
+  }
+
+  override def processRecords(processRecordsInput: ProcessRecordsInput): Unit = {
+    if (processRecordsInput.isAtShardEnd) {
+      semaphore.acquire()
+    }
+
+    // This implementation will try to checkpoint every Record with the original
+    // checkpointer. Other option would be to keep a reference of the latest
+    // checkpointer passed to this instance using any of these methods:
+    // * processRecords
+    // * shutdownRequested
+    // * shardEnded
+    val checkpoint = (record: KinesisClientRecord) =>
+      processRecordsInput.checkpointer().checkpoint(record.sequenceNumber(), record.subSequenceNumber())
```

Comment:
Would that be something the user would want to decide?

Reply:
Do you mean using the "original" checkpointer vs use the "latest" checkpointer?

I think it's a low-level decission that the library should take for the user. It looks to work both ways, although that can change in the future. I'll try to improve the code a bit btw.